src/components/EditorWidgets/Markdown/serializers/slateRemark.js

```diff
@@ -234,6 +234,23 @@ function convertTextNode(node) {
    */
   if (node.leaves) {
     const processedLeaves = node.leaves.map(processLeaves);
+    // Compensate for Slate including leading and trailing whitespace in styled text nodes, which
+    // cannot be represented in markdown (https://github.com/netlify/netlify-cms/issues/1448)
+    for (let i = 0; i < processedLeaves.length; i += 1) {
+      const leaf = processedLeaves[i];
+      if (leaf.marks.length > 0 && leaf.text && leaf.text.trim() !== leaf.text) {
+        const [, leadingWhitespace, trimmedText, trailingWhitespace] = leaf.text.match(/^(\s*)(.*?)(\s*)$/);
+        leaf.text = trimmedText;
+        if (leadingWhitespace.length > 0) {
+          processedLeaves.splice(i, 0, { text: leadingWhitespace, marks: [], textNodeType: leaf.textNodeType });
+          i += 1;
+        }
+        if (trailingWhitespace.length > 0) {
+          processedLeaves.splice(i + 1, 0, { text: trailingWhitespace, marks: [], textNodeType: leaf.textNodeType });
+          i += 1;
+        }
+      }
+    }
```

Comment:
We try to do things in an immutable fashion - I'd expect us to [`flatMap`](https://lodash.com/docs/4.17.10#flatMap) `processedLeaves` with a function that does the following:

```js

if (leaf.marks.length > 0 && trim(leaf.text) !== leaf.text) {
  const [_ , leadingWhitespace, trimmedText, trailingWhitespace] = leaf.text.match(/^(\s*)(.*?)(\s*)$/);
  const leafDefaults = { marks: [], textNodeType: leaf.textNodeType };
  return [
    leadingWhitespace.length > 0 && { ...leafDefaults, text: leadingWhitespace },
    { ...leaf, text: trimmedText },
    trailingWhitespace.length > 0 && { ...leafDefaults, text: trailingWhitespace },
  ];
}
```

We should also see how if upgrading Slate alleviates this issue (it may not).

Reply:
Okay, I'm not really familiar with immutable.js, but I can take a look when I'm back from vacation in a week.

I sincerely doubt that upgrading Slate will have any effect on this. It seems like a netlify-cms bug in the html => markdown conversion when styled text begins or ends with whitespace. It seems unlikely that slate will change its data model in a way that happens to resolve the problem.

Reply:
Agreed. Really glad to see someone braving the waters of our Slate/Remark implementation!

Note, I'm not saying to use Immutable.js, just that we want to do things in an immutable fashion, which is to say we create new values instead of changing existing ones.

Reply:
To be honest I'm not sure exactly how to accomplish that in a performant way. Did you notice that `processedLeaves` is created via a `map` in the same function right before the patching kicks in? https://github.com/netlify/netlify-cms/pull/1517/files#diff-c3642671f69e97062a8b37103f621c5fR219 ... and that `processLeaves` creates a copy of each leaf?

I'm also a fan of immutability as a principle, but I think this (pragmatically) qualifies since `convertTextNode` only mutates values that it has created itself.